reactor-core/src/jcstress/java/reactor/core/publisher/FluxPublishOnStressTest.java

```diff
@@ -0,0 +1,119 @@
+ * limitations under the License.
+ */
+package reactor.core.publisher;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+import org.openjdk.jcstress.annotations.Actor;
+import org.openjdk.jcstress.annotations.Arbiter;
+import org.openjdk.jcstress.annotations.JCStressTest;
+import org.openjdk.jcstress.annotations.Outcome;
+import org.openjdk.jcstress.annotations.State;
+import org.openjdk.jcstress.infra.results.I_Result;
+import org.reactivestreams.Subscription;
+
+import reactor.core.scheduler.Schedulers;
+import reactor.util.concurrent.Queues;
+import reactor.util.context.Context;
+
+import static org.openjdk.jcstress.annotations.Expect.*;
+
+abstract class FluxPublishOnStressTest {
+
+	@JCStressTest
+	//note: -1 can still happen but should be present in a very low number
+	@Outcome(id = {"-1"}, expect = ACCEPTABLE_INTERESTING, desc = "Latch timed out")
+	@Outcome(id = {"1"}, expect = ACCEPTABLE,  desc = "Simple discard")
+	@Outcome(id = {"10"}, expect = ACCEPTABLE,  desc = "Processed")
+	@Outcome(id = {"1001"}, expect = ACCEPTABLE_INTERESTING, desc = "Discarded and onError")
+	@Outcome(id = {"1010"}, expect = ACCEPTABLE_INTERESTING, desc = "Processed and onError")
+	@Outcome(id = {"-2"}, expect = FORBIDDEN, desc = "Latch interrupted")
+	@Outcome(id = {"0"}, expect = FORBIDDEN,  desc = "Nothing happened")
+	@Outcome(id = {"2"}, expect = FORBIDDEN,  desc = "Double discard")
+	@Outcome(id = {"11"}, expect = FORBIDDEN,  desc = "Processed and discarded")
+	@Outcome(id = {"1000"}, expect = FORBIDDEN, desc = "only onError")
+	@Outcome(id = {"1002"}, expect = FORBIDDEN, desc = "double discard and onError")
+	@Outcome(id = {"1011"}, expect = FORBIDDEN, desc = "processed and discarded and onError")
+	@State
+	public static class QueueAndCancelDiscardStressTest extends FluxPublishOnStressTest {
+
+		AtomicInteger           state = new AtomicInteger(0);
+		CountDownLatch latch = new CountDownLatch(2);
+
+		LambdaSubscriber<AtomicInteger> downstream = new LambdaSubscriber<>(
+				ai -> {
+					ai.addAndGet(10);
+					latch.countDown();
+				},
+				t -> {
+					state.addAndGet(1000);
+					latch.countDown();
+				},
+				() -> {}, s -> s.request(1L),
+				Operators.enableOnDiscard(Context.empty(), (Consumer<AtomicInteger>) ai -> {
+					ai.incrementAndGet();
+					latch.countDown();
+				}));
+
+		FluxPublishOn.PublishOnSubscriber<AtomicInteger> publishOnSubscriber = new FluxPublishOn.PublishOnSubscriber<>(
+				downstream,
+				Schedulers.immediate(), Schedulers.immediate().createWorker(), false, 10, 10,
+				Queues.small());
+
+
+		{
+			publishOnSubscriber.onSubscribe(new Subscription() {
+				@Override
+				public void request(long n) {
+
+				}
+
+				@Override
+				public void cancel() {
+				}
+			});
+		}
+
+		@Actor
+		void inner() {
+			publishOnSubscriber.onNext(state);
```

Comment:
can we add a scenario of sending 2 onNext. I guess it may result in an element never dropped from the queue